File: backend/src/main/kotlin/net/animeta/backend/chart/SundayStartWeekRange.kt
package net.animeta.backend.chart

import java.time.DayOfWeek
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.TemporalAdjusters

data class SundayStartWeekRange(val sunday: LocalDate) : ChartRange {
    companion object {
        fun now(zoneId: ZoneId): SundayStartWeekRange {
            return including(LocalDate.now(zoneId))
        }

        fun including(date: LocalDate): SundayStartWeekRange {
            return SundayStartWeekRange(date.with(TemporalAdjusters.previousOrSame(DayOfWeek.SUNDAY)))
        }
    }

    override fun prev(): SundayStartWeekRange {
        return SundayStartWeekRange(sunday.minusWeeks(1))
    }

    override fun startDate(): LocalDate {
        return sunday
    }

    override fun endDate(): LocalDate {
        return sunday.plusWeeks(1).minusDays(1)
    }
}

File: backend-ts/src/utils/sunday_start_week_range.ts
import { Temporal } from '@js-temporal/polyfill';

export class SundayStartWeekRange {
  constructor(public sunday: Temporal.PlainDate) {}

  equals(other: SundayStartWeekRange): boolean {
    return this.sunday.equals(other.sunday);
  }

  prev(): SundayStartWeekRange {
    return new SundayStartWeekRange(this.sunday.subtract({ weeks: 1 }))
  }

  startDate(): Temporal.PlainDate {
    return this.sunday
  }

  endDate(): Temporal.PlainDate {
    return this.sunday.add({ weeks: 1 }).subtract({ days: 1 })
  }

  static now(timeZone: Temporal.TimeZone | Temporal.TimeZoneProtocol): SundayStartWeekRange {
    return this.including(Temporal.Now.plainDateISO(timeZone))
  }

  static including(date: Temporal.PlainDate): SundayStartWeekRange {
    if (date.dayOfWeek === 7)
      return new SundayStartWeekRange(date)
    return new SundayStartWeekRange(date.subtract({ days: date.dayOfWeek }))
  }
}